### PR TITLE
Extend support of `_` in type parameters

### LINKED ISSUE #67995: ICE: Returning impl Trait with placeholder type parameter
Hope this isn't a duplicate.

**EDIT: Updated example, removing async/await**

### Minimal Example
```rust
fn main() {
    evil();
}

trait Trait<T> {}

fn evil() -> impl Trait<_> {}
```

### Workaround

Avoid placeholders.
`fn evil() -> impl Trait<()> {`

### Meta

`rustc 1.42.0-nightly (859764425 2020-01-07) running on x86_64-unknown-linux-gnu`

EDIT: No ICE on stable Rust 1.40.0, build fails with error `E0121` at `fn evil() ...`.

### Compiler Log

<details>
<summary>Expand</summary>

```
    Checking crash_test v0.1.0 (/home/richie/prj/crash_test)
error: internal compiler error: bad placeholder type
 --> src/main.rs:7:25
  |
7 | fn evil() -> impl Trait<_> {}
  |                         ^

error: internal compiler error: errors selecting obligation during MIR typeck: [FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<() as Trait<[type error]>>)), depth=0),Ambiguity)]

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:347:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (859764425 2020-01-07) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: could not compile `crash_test`.

To learn more, run the command again with --verbose.
```

</details>

### PR DESCRIPTION
**Pull request by estebank**
- Account for `impl Trait<_>`.
 - Provide a reasonable `Span` for empty `Generics` in `impl`s.
 - Account for `fn foo<_>(_: _) {}` to suggest `fn foo<T>(_: T) {}`.
 - Fix #67995. Follow up to #67597.

### CODE REVIEW

File: src/librustc_typeck/collect.rs [hunk 1]
```diff
@@ -150,12 +150,24 @@ crate fn placeholder_type_error(
...
     if generics.is_empty() {
-        sugg.push((ident_span.shrink_to_hi(), format!("<{}>", type_name)));
+        sugg.push((span, format!("<{}>", type_name)));
     } else {
-        sugg.push((
-            generics.iter().last().unwrap().span.shrink_to_hi(),
-            format!(", {}", type_name),
-        ));
+        if let Some(arg) = generics.iter().find(|arg| match arg.name {
+            hir::ParamName::Plain(ident) if ident.name == kw::Underscore => true,
```
Comment (Centril) [inline src/librustc_typeck/collect.rs]:
```suggestion
            hir::ParamName::Plain(Ident { name: kw::Underscore, .. }) => true,
```

File: src/librustc_typeck/collect.rs [hunk 2]
```diff
@@ -150,12 +150,24 @@ crate fn placeholder_type_error(
...
         placeholder_types.iter().map(|sp| (*sp, type_name.to_string())).collect();
     if generics.is_empty() {
-        sugg.push((ident_span.shrink_to_hi(), format!("<{}>", type_name)));
+        sugg.push((span, format!("<{}>", type_name)));
     } else {
-        sugg.push((
-            generics.iter().last().unwrap().span.shrink_to_hi(),
-            format!(", {}", type_name),
-        ));
+        if let Some(arg) = generics.iter().find(|arg| match arg.name {
```
Comment (Centril) [inline src/librustc_typeck/collect.rs]: `else if let ...`?

Comment (estebank) [reply src/librustc_typeck/collect.rs]: I feel that `if let`/`else` can be harder to read for cases like these:

```
        if let Some(arg) = generics.iter().find(|arg| match arg.name {
            hir::ParamName::Plain(Ident { name: kw::Underscore, .. }) => true,
            _ => false,
        }) {
```

vs

```
        if let Some(arg) = generics.iter().find(|arg| {
            if let hir::ParamName::Plain(Ident { name: kw::Underscore, .. }) = arg.name {
                true
            } else {
                false
            }
        }) {
```

Comment (estebank) [reply src/librustc_typeck/collect.rs]: ah, I see what you meant

File: src/librustc_typeck/collect.rs [hunk 3]
```diff
@@ -1789,10 +1800,19 @@ fn find_opaque_ty_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
...
 fn is_suggestable_infer_ty(ty: &hir::Ty<'_>) -> bool {
+    use hir::TyKind::*;
     match &ty.kind {
-        hir::TyKind::Infer => true,
-        hir::TyKind::Slice(ty) | hir::TyKind::Array(ty, _) => is_suggestable_infer_ty(ty),
-        hir::TyKind::Tup(tys) => tys.iter().any(|ty| is_suggestable_infer_ty(ty)),
+        Infer => true,
+        Slice(ty) | Array(ty, _) => is_suggestable_infer_ty(ty),
+        Tup(tys) => tys.iter().any(is_suggestable_infer_ty),
+        Ptr(mut_ty) | Rptr(_, mut_ty) => is_suggestable_infer_ty(mut_ty.ty),
```
Comment (Centril) [inline src/librustc_typeck/collect.rs]: Let's exercise both cases in the tests.

File: src/librustc_typeck/collect.rs [hunk 4]
```diff
@@ -172,16 +179,20 @@ fn reject_placeholder_type_signatures_in_item(tcx: TyCtxt<'tcx>, item: &'tcx hir
     let (generics, suggest) = match &item.kind {
         hir::ItemKind::Union(_, generics)
         | hir::ItemKind::Enum(_, generics)
-        | hir::ItemKind::Struct(_, generics) => (&generics.params[..], true),
-        hir::ItemKind::TyAlias(_, generics) => (&generics.params[..], false),
+        | hir::ItemKind::TraitAlias(generics, _)
+        | hir::ItemKind::Trait(_, _, generics, ..)
+        | hir::ItemKind::Impl(_, _, _, generics, ..)
+        | hir::ItemKind::Struct(_, generics) => (generics, true),
+        hir::ItemKind::OpaqueTy(hir::OpaqueTy { generics, .. })
```
Comment (Centril) [inline src/librustc_typeck/collect.rs]: Let's exercise trait aliases and opaque types in tests.

**Comment by Centril** [comment]
r=me with ^-- addressed + comments in private

**Pull request merged**